core/src/main/scala/aima/core/environment/vacuum/VacuumEnvironment.scala

```diff
@@ -6,34 +6,14 @@ import aima.core.random.DefaultRandomness
 /**
   * @author Shawn Garner
   */
-case class VacuumEnvironment(map: VacuumMap = VacuumMap()) extends Environment[VacuumAction, VacuumPercept] { self =>
+case class Vacuum(map: VacuumMap = VacuumMap()) extends Environment[Vacuum, VacuumPercept, VacuumAction] {
 
-  def addAgent(agent: Agent[VacuumAction, VacuumPercept]): Environment[VacuumAction, VacuumPercept] = {
-    VacuumEnvironment(map.addAgent(agent))
+  def addAgent(agent: Agent[Vacuum, VacuumPercept, VacuumAction]): Vacuum = {
+    Vacuum(map.addAgent(agent))
   }
 
-  def removeAgent(agent: Agent[VacuumAction, VacuumPercept]): Environment[VacuumAction, VacuumPercept] = {
-    VacuumEnvironment(map.removeAgent(agent))
-  }
-
-  def actuate(
```

Comment:
@benderpremier This is the part I don't like.  We just blindly trust the agent to do things to the environment and could lead to abuses.  IMO the environment dictates the rules and the agent can only act through the rules the environment dictates.  I also like this in that if you black list an agent you can control it's actions after it's been added.

Reply:
@benderpremier the other thought I have is what if you have a read only agent but someone hacks it to change the environment that does more than let it just perceive the environment whereas other agents are participants and can do certain actions, and other agents can perform more actions.  I'm thinking of privilege escalation problems.

Reply:
Yeah I understand your argument. As I said above I think if you can decide to either decide to trust only the environment or both the argument and the actuators.

I guess if we were designing a tool to benchmark an agent performance, the only part that you'd allow the user of the library to change would be the agent function. So I guess you'd want to make the Environment, Actuator and Sensor trait private and only allow the user to provide an agent function.